### PR TITLE
	Move part of the SymbolFinder out of proc.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
This PR makes it so that for "declaration finding" APIs are now able to run out of proc.  These APIs are primarily used by Add-Using and Full-Qualify to find declarations with names that match what the user has written.

Doing the search out of proc means that the remote process can build its own indices here (using hte incremental analyzer infrastructure), and costly computations will not generate lots of garbage in the host process.

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs [hunk 1]
```diff
@@ -115,5 +131,40 @@ internal static partial class DeclarationFinder
...
+                return (true, rehydrated);
+            }
+
+            return (false, ImmutableArray<SymbolAndProjectId>.Empty);
+        }
+
+        private static async Task<ImmutableArray<SymbolAndProjectId>> RehydrateAsync(
+            Solution solution, SerializableSymbolAndProjectId[] array, CancellationToken cancellationToken)
+        {
+            var result = ArrayBuilder<SymbolAndProjectId>.GetInstance();
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs]: Is there a way to pre-allocate the size, given `array.Length`?

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_SourceDeclarations.cs [hunk 2]
```diff
@@ -17,7 +17,12 @@ namespace Microsoft.CodeAnalysis.FindSymbols
 
     internal static partial class DeclarationFinder
     {
-        internal static async Task<ImmutableArray<SymbolAndProjectId>> FindSourceDeclarationsWithNormalQueryAsync(
+        #region Dispatch Members
+
+        // These are the public entrypoints to finding source declarations.  They will attempt to
+        // remove the query to the OOP process, and will fallback to local processing if they can't.
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_SourceDeclarations.cs]: s/remove/remote/.

Consider not using OOP as an acronym.

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs [hunk 3]
```diff
@@ -18,52 +16,12 @@ namespace Microsoft.CodeAnalysis.FindSymbols
...
-        /// </summary>
-        internal static async Task<ImmutableArray<SymbolAndProjectId>> FindAllDeclarationsAsync(
-            Project project, string name, bool ignoreCase, SymbolFilter filter, CancellationToken cancellationToken = default(CancellationToken))
-        {
-            if (name == null)
-            {
-                throw new ArgumentNullException(nameof(name));
-            }
-
-            if (string.IsNullOrWhiteSpace(name))
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs]: Did this change lose the handling of empty names, or does that also exist at a lower level?

Is it important?

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs]: Sure.

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs]: That check is preserved in: FindAllDeclarationsWithNormalQueryAsync

**Pull request merged**